### ISSUE TITLE
Matrix multiplication not supported

### ISSUE THREAD

**Issue by amakropoulos** [Unity-Technologies/barracuda-release#68]
Hi all, thanks for this great project!

I am trying to do matrix multiplication but it doesn't seem to be supported in Barracuda.
Do you have any suggestions?

The code I am using (in pytorch) is:
```python
class Multiply(torch.nn.Module):
    def forward(self, x, y):
        return torch.matmul(x, y)
```
and to export in ONNX:
```python
model = Multiply()
model_input = (
    torch.rand([1, 2, 10, 5]),
    torch.rand([1, 2, 5, 10])
)
torch.onnx.export(
    model, model_input, "multiply.onnx",
    do_constant_folding=True, verbose=True, export_params=True,
    opset_version=9
)
```
I attach the exported ONNX file:
[multiply.onnx.zip](https://github.com/Unity-Technologies/barracuda-release/files/4835944/multiply.onnx.zip)

The error I get when importing the asset in Unity is the following:
```
OnnxImportException: Unexpected error while parsing layer 2 of type MatMul.
Currently only constant tensors are supported for `B` input in node of type MatMul. Instead 2.B is pointing to non constant node 1.

Json: { "input": [ "0", "1" ], "output": [ "2" ], "name": "MatMul_0", "opType": "MatMul" }
  at Unity.Barracuda.ONNXNodeWrapper.GetRequiredInputAsConstant (System.String input, System.String onnxLayout, System.String onnxName) [0x00035] in /home/benuix/Barracuda/Library/PackageCache/com.unity.barracuda@1.0.0/Barracuda/Editor/ONNXNodeWrapper.cs:203 
  at Unity.Barracuda.ONNXNodeWrapper.Input1Constant (System.String onnxLayout, System.String name) [0x00001] in /home/benuix/Barracuda/Library/PackageCache/com.unity.barracuda@1.0.0/Barracuda/Editor/ONNXNodeWrapper.cs:70 
  at Unity.Barracuda.ONNXModelImporter.<.ctor>b__14_59 (Unity.Barracuda.ModelBuilder net, Unity.Barracuda.ONNXNodeWrapper node) [0x00001] in /home/benuix/Barracuda/Library/PackageCache/com.unity.barracuda@1.0.0/Barracuda/Editor/ONNXModelImporter.cs:623 
  at Unity.Barracuda.ONNXModelImporter.ConvertOnnxModel (Onnx.ModelProto onnxModel) [0x00367] in /home/benuix/Barracuda/Library/PackageCache/com.unity.barracuda@1.0.0/Barracuda/Editor/ONNXModelImporter.cs:1083 

Unity.Barracuda.ONNXModelImporter.Err (Unity.Barracuda.Model model, System.String layerName, System.String message, System.String extendedMessage, System.String debugMessage) (at Library/PackageCache/com.unity.barracuda@1.0.0/Barracuda/Editor/ONNXModelImporter.cs:1399)
Unity.Barracuda.ONNXModelImporter.ConvertOnnxModel (Onnx.ModelProto onnxModel) (at Library/PackageCache/com.unity.barracuda@1.0.0/Barracuda/Editor/ONNXModelImporter.cs:1092)
Unity.Barracuda.ONNXModelImporter.OnImportAsset (UnityEditor.Experimental.AssetImporters.AssetImportContext ctx) (at Library/PackageCache/com.unity.barracuda@1.0.0/Barracuda/Editor/ONNXModelImporter.cs:1000)
UnityEditor.Experimental.AssetImporters.ScriptedImporter.GenerateAssetData (UnityEditor.Experimental.AssetImporters.AssetImportContext ctx) (at /home/builduser/buildslave/unity/build/Modules/AssetPipelineEditor/Public/ScriptedImporter.cs:22)
```

**Comment by AlexRibard** [comment]
Hi!
Onnx exports Linear layers at MatMul (X * W + B) and in this case the both W and B are constants.
For optimization purposes, we optimized for this case, assuming that W and B are constants.
So we currently don't accept varying W and B. This will be fixed.

In the meantime, you can still run the MatMul, but you'll need to do the following:
```
private readonly PrecompiledComputeOps m_Ops = new PrecompiledComputeOps(); // if you want GPU ops
private readonly BurstCPUOps m_Ops = new BurstCPUOps(); // for CPU
...
Tensor Y= m_Ops.MatMul(X, false, W, false);
```

**Comment by amakropoulos** [comment]
Hi Alex, thank you for the quick reply!

This indeed solves the problem of needing to have the second matrix as a constant.
There are some new problems though that are created.

Do you have any suggestions?
Thank you again!

**GPU**
MatMul still calls Dense under the hood:
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Barracuda/Runtime/Core/Backends/BarracudaReferenceCompute.cs#L822
Dense only allows the second matrix (W) to have a dimensionality of 2:
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Barracuda/Runtime/Core/Backends/BarracudaCompute.cs#L809
One solution is to iterate through batch and channel dimension and do MatMul for the submatrices (see solution below).
However this fails at this assertion:
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Barracuda/Runtime/Core/Backends/BarracudaCompute.cs#L828
If I comment the assertion out, I get a result which is actually wrong (see below).

**CPU**
This doesn't use Dense but requires both matrices to have a dimensionality of 2:
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Barracuda/Runtime/Core/Backends/BarracudaBurstCPU.Ops.cs#L28-L29
If I again do the iteration solution it returns the correct result.

**Hacky solution**
A hacky solution is to iterate through batch and channel dimension and do MatMul for the submatrices.
This returns the expected result for CPU (an array full of 20) but not for the GPU (an array full of 22).
```C#
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Unity.Barracuda;

public class Test : MonoBehaviour
{
    private IOps m_Ops;

    private void MatMulForBatchChannel(Tensor input1, Tensor input2, ref Tensor output, int b, int c){
        var shape = input1.shape.ToArray();
        int height = shape[1];
        int width = shape[2];

        Tensor part1 = new Tensor(height, width);
        Tensor part2 = new Tensor(width, height);
        for (var h = 0; h < height; h++)
            for (var w = 0; w < width; w++){
                part1[h, w] = input1[b, h, w, c];
                part2[w, h] = input2[b, w, h, c];
            }

        Tensor outputpart = m_Ops.MatMul(part1, false, part2, false);
        for (var h = 0; h < height; h++)
            for (var w = 0; w < height; w++)
                output[b, h, w, c] = outputpart[h, w];
    }

    private void initInput(ref Tensor input, float value){
        var shape = input.shape.ToArray();
        for (var b = 0; b < shape[0]; b++)
            for (var c = 0; c < shape[3]; c++)
                for (var h = 0; h < shape[1]; h++)
                    for (var w = 0; w < shape[2]; w++)
                        input[b, h, w, c] = value;
    }

    private void printOutput(ref Tensor output){
        string result = "";
        var shape = output.shape.ToArray();
        for (var b = 0; b < shape[0]; b++)
            for (var c = 0; c < shape[3]; c++)
                for (var h = 0; h < shape[1]; h++){
                    for (var w = 0; w < shape[2]; w++){
                        result += output[b, h, w, c].ToString();
                        if (w < shape[2]-1)
                            result += ',';
                    }
                    result += "\n";
                }
        Debug.Log(shape);
        Debug.Log(result);
    }

    void Start()
    {
        // GPU    
        // m_Ops = new PrecompiledComputeOps(ComputeShaderSingleton.Instance.kernels,
        //                                     ComputeShaderSingleton.Instance.referenceKernels, null, false);

        // CPU
        m_Ops = new BurstCPUOps(null);

        int batch = 1;
        int height = 3;
        int width = 10;
        int channels = 2;

        Tensor input1 = new Tensor(batch, height, width, channels);
        Tensor input2 = new Tensor(batch, width, height, channels);
        Tensor output = new Tensor(batch, height, height, channels);

        initInput(ref input1, 1);
        initInput(ref input2, 2);

        for (var b = 0; b < batch; b++)
            for (var c = 0; c < channels; c++)
                MatMulForBatchChannel(input1, input2, ref output, b, c);

        printOutput(ref output);

        m_Ops.MatMul(input1, false, input2, false);

        input1.Dispose();
        input2.Dispose();
        output.Dispose();
    }

    void Update()
    {

    }
}
```

**Comment by AlexRibard** [comment]
Ah yes. You found out what I was referring to (Dense being optimized for const weights and biases) in the GPU case.
What is the issue with the dimensionality?
```
    public int dimensions { get {
            return
                (batch > 1 ? 1 : 0) +
                (height > 1 ? 1 : 0) +
                (width > 1 ? 1 : 0) +
                (channels > 1 ? 1 : 0);
        }
    }
```
So if your matrices are [1,H,W,1] it's fine.
But I see that you are trying with [1, 2, 10, 5] (NCHW) so [1, 10, 5, 2] NHWC
Let me see what the docs say about that behaviour. Does it do 2 MatMul of [10,5] ?

**Comment by AlexRibard** [comment]
Ok apparently that seems to be the behaviour:
2 MatMul [10,5] x [5, 10]

So you can loop over #channels unpack matrices, do the matmul, re-pack in output

**Comment by amakropoulos** [comment]
Perfect thank you I'll do that!

Before closing this, do you have any ideas about the other issues with GPU?

> GPU fails at this assertion:
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Barracuda/Runtime/Core/Backends/BarracudaCompute.cs#L828
If I comment the assertion out, I get a result which is actually wrong (see below).

> This returns the expected result for CPU (an array full of 20) but not for the GPU (an array full of 22).

**Comment by AlexRibard** [comment]
Erratum, just looked at your code. You are doing exactly that! :)
As for the GPU code. Using anything else than `ReferenceCompute` will lead to invalid result.
The problem is we pack W and B together as a single buffer (WBK buffer).
Just commenting out the Assert will not work as we will be reading oom and thus invalid data.
If you _really_ want to make it work on GPU, it gets more tricky as you need to create the WBK tensor yourself.
So W-followed by B (0 in this case).
Check out the code in `BarracudaCompute.Conv2DTrans`:
```
ComputeBuffer buffer = new ComputeBuffer(W.shape.length + B.shape.length, sizeof(float));

var W = new Tensor(K.shape, new SharedComputeTensorData(buffer, W.shape, 0));
var B = new Tensor(B.shape, new SharedComputeTensorData(buffer, B.shape, W.shape.length));
...
Dispatch compute shader to fill in the buffer
```
I can chat with you via email if you want more detail on this

**Comment by amakropoulos** [comment]
yes please that would be great! 
my email is benuix AT gmail DOT com.

I'm closing the issue, thank you for all the help!

**Comment by albanmdb3d** [comment]
Hi, sorry to reopen this issue, but I'm also interested in doing matrix multiplications on the GPU. Would it be possible to share your solution if you found one?

**Comment by AlexRibard** [comment]
Hi!
Yes, that got pushed in latest. It will be pushed in 1.1
Here is the code if you need it urgently

BarracudaCompute.cs:
```
    public override Tensor MatMul(Tensor X, bool xTranspose, Tensor Y, bool yTranspose)
    {
        // MatMul implementation in terms of Dense
        var A = (xTranspose) ? Transpose(X): X;
        var B = (yTranspose) ? Transpose(Y): Y;
        var Cshape = new TensorShape(1, B.flatWidth); // intialize bias with zeros

        ComputeBuffer buffer = new ComputeBuffer(B.shape.length + Cshape.length, sizeof(float));

        var Bpacked = new Tensor(B.shape, new SharedComputeTensorData(buffer, B.shape, 0));
        var Cpacked = new Tensor(Cshape, new SharedComputeTensorData(buffer, Cshape, B.shape.length));

        var fn_pack = new ComputeKernel(new ComputeFunc(m_Kernels, "MatMulPackB0Bias"), (B.flatWidth, B.flatHeight, 1));
        fn_pack.SetTensor("X", B.shape, Pin(B).buffer);
        fn_pack.SetTensor("O", Bpacked.shape, Pin(Bpacked).buffer);

        fn_pack.Dispatch();

        var O = Dense(A, Bpacked, Cpacked, Layer.FusedActivation.None);
        if (A != X) A.Dispose();
        if (B != Y) B.Dispose();

        buffer.Dispose();

        return O;
    }
```
And the shader code to pack B and C (in this case I assume 0s for C as ONNX doc for MatMul is only A*B)

Dense.compute
```
#pragma kernel MatMulPackB0Bias

[numthreads(8, 8, 1)]
void MatMulPackB0Bias(uint3 dispatchThreadID : SV_DispatchThreadID)
{
    TENSOR_ARGS2(X, O);

    uint x = dispatchThreadID.x;
    uint y = dispatchThreadID.y;

    if (x >= O.GetFlatWidth()) return;
    if (y >= O.GetFlatHeight()) return;

    float v = X.Get(y, x);
    O.Set(y, x, v);
    O.FastSet(O.GetLength() + x, 0.0f);
}
```

**Comment by mohamedhassanmus** [comment]
Hi @AlexRibard 
I tried to test this in Barracuda 1.1.0. I tested the exact same code that @amakropoulos wrote. Basically multiplying two numbers.

```
class Multiply(torch.nn.Module):
    def forward(self, x, y):
        return torch.matmul(x, y)
onnx_model_path = "multiply.onnx"
model = Multiply()
model_input = (
    torch.rand([1, 2, 10, 5]),
    torch.rand([1, 2, 5, 10])
)
torch_out = model(model_input[0], model_input[1])
print('torch_out shape: {}'.format(torch_out.shape))
torch.onnx.export(
    model, model_input, onnx_model_path,
    do_constant_folding=True, verbose=True, export_params=True,
    opset_version=9
)
import onnx

onnx_model = onnx.load(onnx_model_path)
onnx.checker.check_model(onnx_model)
print(onnx.helper.printable_graph(onnx_model.graph))
```
Both onnx and torch output are of the expected shape `[1, 2, 10, 10]`. But when I upload the onnx model to Unity, I see the output has a wrong size of `[1, 10, 5, 2]`. As you can see in the picture below:
![image](https://user-images.githubusercontent.com/32509797/92111596-d59b2200-edec-11ea-9690-e334e5b54905.png).

The onnx model 
[multiply.onnx.zip](https://github.com/Unity-Technologies/barracuda-release/files/5168309/multiply.onnx.zip)

Any ideas?
Thanks a lot in advance

**Comment by AlexRibard** [comment]
Hi @MohameHassan !
The code I mentioned above is included in 1.1 so you should be able to run your model as is.

The layout change is to be expected, refer to our doc for details
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Documentation%7E/TensorHandling.md

ONNX and pytroch are using a NCHW internal data layout, whereas us follow Tensorflow NHWC layout.
So your `[1,2,10,10]` (N=1, C=2, H=10, W=10) becomes transposed into (N=1, H=10, W=10, C=2) so `[1,10,10,2]` in Unity.
So you'll need to transpose (`np.transpose(0,2,3,1)`) your input accordingly if you want to feed it to Unity

Hope it clarifies.

**Comment by mohamedhassanmus** [comment]
Hi @AlexRibard,
Thanks a lot for the quick response. I am actually still confused though, maybe I am missing something here.
I understand the difference in the layout between ONNX/PyTorch and Barracuda. So I understand that my first PyTorch tensor of `[1,2,10,5]` will become `[1,10,5,2]`. Similarly my second PyTorch tensor of `[1,2,5,10]` will become `[1,5,10,2]`. This is fine. 

The problem, however, is that I expected my output in Barracuda to be of the size `[1,10,10,2]`, but the output in Barracuda is shown to be `[1,10,5,2]`. So I guess there is a difference in the operation itself or I don't get the expected behavior.

**Comment by AlexRibard** [comment]
Ah! Apologies, I read your post in diagonal and missed the main issue!
You are 100% right, the shapes are messed up.
I'll debug it asap!
Thanks for pointing it out

**Comment by AlexRibard** [comment]
@MohameHassan Multi dimension MatMul is not supported atm. Only 2D MatMul is working.
I am working on support, but atm you'll need to iterate over channels and feed those 2D Tensors to MatMul and Concat back at the end.

**Comment by mohamedhassanmus** [comment]
Hi @AlexRibard. Thanks for following-up.
I just tried it 2D MatMul case but it still fails.
```
class Multiply(torch.nn.Module):
    def forward(self, x, y):
        return torch.matmul(x, y)
onnx_model_path = "multiply.onnx"
model = Multiply()
model_input = (
    torch.rand([10, 5]),
    torch.rand([5, 10])
)
input_names = ["x", 'y']
torch_out = model(model_input[0], model_input[1])
print('torch_out shape: {}'.format(torch_out.shape))
torch.onnx.export(
    model, model_input, onnx_model_path, input_names=input_names,
    do_constant_folding=True, verbose=True, export_params=True,
    opset_version=9
)
```

The expected output is of size `(10,10)` but What I get `(10,5)`
![image](https://user-images.githubusercontent.com/32509797/93713983-d7404600-fb5f-11ea-8919-453a330287d2.png)

I tried version `1.1.0` as well as `1.1.1`. 
I also tried to feed inputs of the shape `NCHW` with `N=C=1` so
```
model_input = (
    torch.rand([1,1, 10, 5]),
    torch.rand([1,1, 5, 10])
)
```
But the result is the same.

**Comment by AlexRibard** [comment]
@MohameHassan good spot. There is a bug in our UI, the displayed shape is wrong.
However, the code works (we have multiple test validating comparing with onnxruntime). 
You can validate by executing your network and checking the output.

I will fix the display bug.
On a side note I have implemented dynamic multi-dim matmul. It will be pushed in the next release.
If you want the code early email me at barracuda-support@unity3d.com I'll send you the PR.
